modules/libcom/src/fdmgr/fdManager.cpp

```diff
@@ -19,122 +19,256 @@
 // 1) This library is not thread safe
 //
 
-#include <algorithm>
-
 #define instantiateRecourceLib
 #include "epicsAssert.h"
 #include "epicsThread.h"
 #include "fdManager.h"
 #include "locationException.h"
 
-using std :: max;
+#if !defined(FDMGR_USE_POLL) && !defined(FDMGR_USE_SELECT)
+#if defined(__linux__) || defined(darwin) || _WIN32_WINNT >= 0x600 || (defined(__rtems__) && !defined(RTEMS_LEGACY_STACK))
+#define FDMGR_USE_POLL
+#else
+#define FDMGR_USE_SELECT
+#endif
+#endif
+
+#ifdef FDMGR_USE_POLL
+#include <vector>
+#if !defined(_WIN32)
+#include <poll.h>
+#endif
+
+static const short PollEvents[] = { // must match fdRegType
+    POLLRDNORM | POLLRDBAND | POLLIN | POLLHUP | POLLERR,
+    POLLWRBAND | POLLWRNORM | POLLOUT | POLLERR,
+    POLLPRI };
+
+#if defined(_WIN32)
+#define poll WSAPoll
+// Filter out PollEvents that Windows does not accept in events (only returns in revents)
+#define WIN_POLLEVENT_FILTER(ev) static_cast<short>((ev) & (POLLIN | POLLOUT))
+#else
+// Linux, MacOS and RTEMS don't care
+#define WIN_POLLEVENT_FILTER(ev) (ev)
+#endif
+#endif
+
+#ifdef FDMGR_USE_SELECT
+#include <algorithm>
+#endif
+
+struct fdManagerPrivate {
+    tsDLList<fdReg> regList;
+    tsDLList<fdReg> activeList;
+    resTable<fdReg, fdRegId> fdTbl;
+    const double sleepQuantum;
+    epics::auto_ptr<epicsTimerQueuePassive> pTimerQueue;
+    bool processInProg;
+
+#ifdef FDMGR_USE_POLL
+    std::vector<struct pollfd> pollfds;
+#endif
+
+#ifdef FDMGR_USE_SELECT
+    fd_set fdSets[fdrNEnums];
+    SOCKET maxFD;
+#endif
+
+    //
+    // Set to fdreg when in call back
+    // and nill otherwise
+    //
+    volatile fdReg* pCBReg;
```

Comment:
What is the intent of adding `volatile` here?

I don't think it would have any practical effect.  fdManager has never been reentrant.

Reply:
To silence a Codacy warning. It did not believe that calling a callback function could change the value.

Reply:
Well, codacy can not know that fdmanager is not designed to be reentrant, and having `volatile` here doesn't hurt anything.  So ok.